src/services/persistent-storage.service.ts

```diff
@@ -0,0 +1,67 @@
+import { Promisify } from "../utils/types";
+import { defer } from "../utils/HelperFunctions";
+
+export class PersistentStorage implements Promisify<Storage> {
+  private static instances = new Map<string, PersistentStorage>();
+
+  private readonly map = new Map<string, string>();
+
+  constructor(
+    private readonly prefix: string,
+    private readonly storage: Storage = window?.localStorage
+  ) {
+    if (!storage) {
+      throw new Error("No storage available");
+    }
+
+    if (PersistentStorage.instances.has(prefix)) {
+      return PersistentStorage.instances.get(prefix)!;
+    }
+    PersistentStorage.instances.set(prefix, this);
+    this.init();
+  }
+
+  private init() {
+    Array(this.storage.length)
+      .fill(null)
+      .map((_, index) => {
+        return this.storage.key(index)!;
+      })
+      .filter((key) => key?.startsWith(this.prefix))
+      .forEach((key) => {
+        this.map.set(key.slice(this.prefix.length), this.storage.getItem(key)!);
+      });
+  }
```

Comment:
That will not really change the fact that the underscore rule will have to be added at some point in the future for any type of callback functions.  As for an imperative loop, sure, we could do that but I'm quite not fond of the syntax of a `for(let i = 0` as I find that sometimes this encourages developers to use only imperative loops and end up not understanding the array methods provided by js.

As for the second example, that's how I wrote it at the beginning. The only problem is that `Object.keys(storage)` will not give the same response as storage can be a `Storage` type like so it might not behave the same way (e.g: tests and mocking the `localStorage`).

Reply:
Not sure what you mean by needing it for any callback function in the future? While arity can be a problem in JavaScript, I've not found a case where another approach can't be used instead. In the existing code we have a lot of additional work for is effectively a loop based on the number of keys in local storage.

My priority on these sorts of problems are less about performance and more about readability (I think it's safe to say a for loop is simpler to understand in this case) but having said that we're currently constructing an n length array, mapping n times, then filtering (to potentially no items) before placing the items into an in-memory Map.

Regarding the second approach - no problem. I would argue that maybe we're thinking too broadly on what should be a relatively simple abstraction. While it's nice that we could support other storage implementations, how likely is it that we'll actually do this **and** it'll support the entire API without problems such as the `.length` functionality

Reply:
I switched to the imperative loop. I agree, it's actually more readable. I would still like to know, however, how do you get around unused variables inside callbacks. Sometimes, depending on the required API, you get some variable that you will not use before the value that's actually required
```
/*
 * This prompts an ESLint error for unused variable
 * The way to hide that error is to prefix it with an _
 */
.someMethod((someUnusedVariables, theActuallyRequiredVariable) => {
   useVariable(theActuallyRequiredVariable);
});
```

Reply:
I believe that's your editor not quite using our ESlint configuration (or at least bundling in some rules of it's own). It's actually happening to my WebStorm editor too. If you run `eslint` from the project manually on the file, you won't get an error. I've not put my attention to linting just yet because we've not fully done the work to have the project in a state where there's no lint problems. Once it hits our GH Actions workflow, we'll soon find what's causing friction

I don't think there's anything wrong with declaring the variable when there's arity issues in a function. All it really means is either the library you're using is badly designed or your use case is an edge case. If it's a module of our own, the same things applies but we can actually do something about it quite easily 😛 A common solution to this problem is the object parameters pattern and is why I merited `near-api-js` for transitioning to this in places